Panache REST guide

gytis opened the pull request:
cc @geoand

Commit:
REST data with Panache guide

gytis commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+
+== Resource customisation
+
+Panache REST provides a `@PanacheRestResource` annotation that can be used to customize certain features of the resource.
```
It's out of scope for this PR but I'm not happy with a name of this annotation (`@PanacheRestResource`). It's very similar to the interface name (`PanacheCrud..Resource`) and might be confusing. Any suggestions on the name change @geoand?

geoand replied:
I dunno.... I'm probably the worst developer when it comes to naming...

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+include::./attributes.adoc[]
+:extension-status: experimental
+
+A lot of web applications are monotonous CRUD applications with boring to write REST APIs.
```
```suggestion
A lot of web applications are monotonous CRUD applications with REST APIs that are tedious to write.
```

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+:extension-status: experimental
+
+A lot of web applications are monotonous CRUD applications with boring to write REST APIs.
+To ease this task Panache REST can generate the basic CRUD endpoints for your entities and repositories.
```
```suggestion
To streamline this task, Panache REST can generate the basic CRUD endpoints for your entities and repositories.
```

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+A lot of web applications are monotonous CRUD applications with boring to write REST APIs.
+To ease this task Panache REST can generate the basic CRUD endpoints for your entities and repositories.
+
+While this extension is still experimental and provides a limited feature set, we hope to get an early feedback for it.
```
Perhaps this should be in a note?

gytis replied:
There is a generic note about the experimental feature already. Do you think I should add another one?

geoand replied:
Oh OK. If it's used elsewhere as well, let's keep it as it is

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+
+Panache REST generates CRUD resources based on the interfaces available in your application.
+For each entity and repository that you want to generate, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
+_Do not implement these interfaces and don't provide custom methods because they won't be implemented._
```
```suggestion
_Do not implement these interfaces and don't provide custom methods because they be ignored._
```

gytis replied:
`will be ignored` maybe?

geoand replied:
yeah :)

gytis replied:
How do you edit string in the PR?

geoand replied:
See `6` from https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+You can, however, override the methods from the extended interface in order to customize them (see the section at the end).
+
+The resources that will be generated will be functionally equivalent for both entities and repositories.
+Only the data will accessed using either one or the other.
```
This line doesn't seem to make sense

gytis replied:
How about `The generated resources will be functionally equivalent for both entities and repositories. Except that the particular data access pattern will be used.`

geoand replied:
My counter: 

```
The generated resources will be functionally equivalent for both entities and repositories., the difference being in the particular data access pattern will be used.
```

gytis replied:
`The only difference being the particular data access pattern in use.`?

geoand replied:
Yeah sorry. I really need to pay more attention to what I actually write as opposed to what I am thinking...

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+Only the data will accessed using either one or the other.
+The skeleton of the generated JAX-RS resource will look like this:
+
+[source,java]
```
I think we need a little more emphasis that on the fact this is generated - not something the user would write.
Mentioning it because looking at the code initially through me off...

gytis replied:
It does say "generated JAX-RS resource". Any suggestion?

geoand replied:
Yeah I know, it's just that it's too easy to think you need to write this. 
Perhaps change the name of the class and add a comment on it?

gytis replied:
I've changed the sentence above

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+
+=== PanacheEntityCrudResource
+
+If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could tell Panache REST to generate its CRUD resource with the following interface:
```
```suggestion
If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+
+=== PanacheRepositoryCrudResource
+
+If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could tell Panache REST to generate its CRUD resource with the following interface:
```
```suggestion
If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could instruct Panache REST to generate its CRUD resource with the following interface:
```

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,148 @@
+
+=== Available options
+
+* `hal` - in addition to the standard `application/json` responses, generates additional methods that can return `application/hal+json` responses if requested via `Accept` header.
```
Perhaps a link to what `hal` is would be nice

gytis replied:
I thought about it, but then though it might be too much info. Do you think it would be useful?

geoand replied:
I think it would be nice to have

gytis replied:
Done

geoand reviewed (commented):
Nice start!

Added some comments

geoand commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,242 @@
+* Add the required dependencies to your `pom.xml`
+** Panache REST with Hiberante ORM extension (`quarkus-panache-rest-hibernate-orm`)
+** A JDBC driver extension (`quarkus-jdbc-postgresql`, `quarkus-jdbc-h2`, `quarkus-jdbc-mariadb`, ...)
+** One of the RestEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
```
```suggestion
** One of the RESTEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
```

gsmet commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,242 @@
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+
```
Please avoid the newline here, the website is very picky about that.

gsmet commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,242 @@
+== Generating CRUD resources
+
+Panache REST generates CRUD resources based on the interfaces available in your application.
+For each entity and repository that you want to generate, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
```
```suggestion
For each entity and repository you want to expose as a REST service, provide an interface that extends either `PanacheEntityCrudResource` or `PanacheRepositoryCrudResource` interface.
```
would be more correct if I understand the extension correctly?

gsmet commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,242 @@
+
+[source,java]
+----
+@Path("/entities") // Base bath is a pluralised name of an entity
```
I think this is a mistake. We shouldn't "pluralize" things as there's a good chance we will get some of them wrong and it really doesn't bring anything.

Keep in mind that a lot of people code in their own languages where the plural rules will be completely different.

Also, frankly, I don't think it's better as a plural so why take the risk?

gsmet replied:
/cc @FroMage

geoand replied:
If we want plural, perhaps let the user specify it?

gsmet replied:
Just to give an example, in French, we have a lot of specific plural forms (`cheval` -> `chevaux` for instance).

gsmet replied:
Yeah, taking the entity name as is would be the default and you could define a specific path in the annotation?

gytis replied:
Would you mind to create an issue for this and assign to me? I'll need to change the extension because it now pluralises everything.
I'll also add a way to override it separately.

Just an FIY Spring Data REST simply adds `s` to every entity name.

gsmet replied:
> Just an FIY Spring Data REST simply adds s to every entity name.

Oh, my. OK I will create an issue.

FroMage replied:
Makes it even more fun with:

```java
@Entity
public class 馬 extends PanacheEntity {
}
```

And the URI ends up being `/馬s`, right? :)

Myself, I've always used `Cheval` for the entity name (singular) and `Chevaux` (plural) for the controller name. And RoR derives the URIs from the controller names, so it would be `/chevaux` correctly.

I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Naturally we should also allow users to override it.

gytis replied:
> And the URI ends up being `/馬s`, right? :)

I would think so. At least for a simple `people` I got a path `/peoples` when i tried it last time.

> I just suggest we advertise the same convention, it solves many issues. And we default to the controller name for the URI, instead of the entity name.

Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.
With an entity we could simply take an its name, make it all lowercase separated with `-` instead of camelcase and get on with it. And then I'll add an option to override the path.

FroMage replied:
> Extracting a name from a controller will be trickier because a user can name his classes differently. We could recommend a convention but who knows if they'll follow it.

The point of conventions is that people follow them. If they don't they can't customise the output.

Entities are already singular by convention. Controller names don't have a convention other than the brain-dead "resource" that jax-rs got confused about.

It's pretty easy to tell people to name their controller `People` and default to using that name.

gytis replied:
Let's discuss it on https://github.com/quarkusio/quarkus/issues/9261 since the guide change won't be enough for this anyway.

gsmet commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,242 @@
+
+=== Available options
+
+* `hal` - in addition to the standard `application/json` responses, generates additional methods that can return `application/hal+json` responses if requested via `Accept` header.
```
Wondering if we should have an array of content types instead with the default being `{ "application/json" }`. That would be much more flexible.

gsmet replied:
/cc @FroMage

geoand replied:
Well, I don't think we currently support any response type, but this seems like a valid point for the future.

gytis replied:
`application/json` is hardcoded and `application/hal+json` can be enabled as an extra. Because HAL is a complex format I had to implement a custom serialiser for it which serialisers special wrapper objects.

gsmet replied:
OK, let's keep it that way for now.

gsmet reviewed (changes requested):
I added some comments, some are related to the implementation itself rather than the doc.

/cc @geoand for further discussions.

gytis commented:
I think I've addresses all the comments except the one about plural paths. For that I'll make a separate guide modification once I change the extension code.

FroMage commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,241 @@
+</dependencies>
+----
+
+* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Panache with Hibernate ORM guide].
```
```suggestion
* Implement the Panache entities and/or repositories as explained in the link:hibernate-orm-panache[Hibernate ORM with Panache guide].
```

FroMage commented on docs/src/main/asciidoc/panache-rest.adoc:
```diff
@@ -0,0 +1,241 @@
+To begin with:
+
+* Add the required dependencies to your `pom.xml`
+** Panache REST with Hiberante ORM extension (`quarkus-panache-rest-hibernate-orm`)
```
Now that I read it like that, I've my doubts over the order, because originally it was supposed to be `X with Panache`, which means `X with style`.

This makes it looks like `Panache` is a noun and the subject instead of an adjective to make Hibernate ORM/REST stylish.

Perhaps it should be `Hibernate ORM REST with Panache` or `RESTful Hibernate ORM with Panache` or `REST for Hibernate ORM with Panache`?

FroMage replied:
I guess, similarly, this guide (and the extension names) would better be "REST with Panache" and `rest-panache`?

Except this sort-of implies it applies to REST apps that aren't necessarily about data, and we plan to have a Panache layer for RESTEasy one day…

Perhaps `RESTful data with Panache`?

FroMage replied:
Or `CRUD with Panache`?

gsmet replied:
In any case, we need to make the extension name consistent with the naming rules we define here.

gytis replied:
Should I create a separate issue to agree on the name there?

FroMage replied:
We can, yes, but this will heavily impact this guide.

gytis replied:
I could just update the guide together with the code once we agree on a name.

FroMage replied:
OK fine.

gytis replied:
https://github.com/quarkusio/quarkus/issues/9293

FroMage reviewed (changes requested):
Great guide. Please don't hate me for reconsidering naming before it's too late…

gytis commented:
I've updated the guide to align with the changes from https://github.com/quarkusio/quarkus/pull/9373 but I'll need to rebase after the merge. There will be a conflict in a `quarkus-extension.yaml` file because of a module rename.

geoand commented:
We'll need a review from @FroMage for this one :)

gytis commented:
Closing this PR because I've cherry-picked the commit to https://github.com/quarkusio/quarkus/pull/9373
